docs/template_plugin/src/template_preprocessing.cpp

```diff
@@ -0,0 +1,113 @@
+// Copyright (C) 2021 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include "template_preprocessing.hpp"
+
+#include <ngraph/opsets/opset3.hpp>
+#include <ngraph/pass/manager.hpp>
+#include <ngraph/pattern/op/wrap_type.hpp>
+#include <ngraph/rt_info.hpp>
+
+using namespace ngraph;
+
+NGRAPH_RTTI_DEFINITION(ngraph::pass::AddPreprocessingMatcher, "AddPreprocessingMatcher", 0);
+
+ngraph::pass::AddPreprocessingMatcher::AddPreprocessingMatcher(const InferenceEngine::InputsDataMap & inputInfoMap) {
```

Comment:
@ilyachur could you remind me what we decided about preprocessing? What plugins should use for preprocessing?
But from my point of view we need to provide nG preprocessing transformations for default path (in case if plugin has no efficient engine for preprocessing).

@ilya-lavrenov I would suggest to separate this pass to three transformations. Also I would simplify transformations interface to consume only necessary preprocessing parameters and other IE dependent code I would leave for plugins.

Reply:
> I would suggest to separate this pass to three transformations

How will I them after? E.g. I need to insert meanValue and stdScale. I have 2 independent transformations: meanValueTransformation and stdScaleTransformation. Once meanValueTransformation has finished and stdScaleTransformation is run, the last one does not know where to add stdScale value (right after parameter, which means no meanValue in prepro, but it's wrong; or after meanValue otherwise - ok). 

So, I think AddPreprocessing is a single step and cannot be split. Or we need to insert some flag that meanValue was added and it will be used by stdScale matcher.

But I agree to make it IE-free (as stated in description of PR).

Reply:
At the current moment plugins should use preprocess info.
But we have a task to support preprocessing in the ngraph function.

Reply:
> At the current moment plugins should use preprocess info.
But we have a task to support preprocessing in the ngraph function.

And what is the conclusion?) Are you OK to generalize this transformation to include to common transformation API?

Reply:
First of all I didn't see support of Color conversion and resize pre-processing. Does template plugin support such pre-processing?

I am OK to generalize this transformation, but plugins should be able to disable this transformation and use GAPI for example.

Reply:
> First of all I didn't see support of Color conversion and resize pre-processing. Does template plugin support such pre-processing?

Color conversion and resize (as well as layout or precision conversion):
1. We don't have nodes to convert them to ngraph operations.
2. It's handled by `inference_engine_preproc` plugin which can convert layouts, precisions, perform color conversion and resize. 

> I am OK to generalize this transformation, but plugins should be able to disable this transformation and use GAPI for example.

From the previous comment we can see that GAPI preproc and current transformation capabilities are NOT overlapped (otherwise I don't need to implement this transformation). This is because most of plugins can fuse scale, mean to the next operations and they do it on their side.

Reply:
Split into 2 transformations to add scale, mean values / images.

The transformation in template plugin still re-used these 2 smaller transformations.